Add Appraisal

deivid-rodriguez opened the pull request:
Closes #3640.
Closes #4510.
Closes #4589.

What do you think about this? The build is passing but I really don't want to bother installing and fighting against Ruby 1.9.3 & Rails 3, that were EOLed maybe about 2 years ago. <strike>So I dropped support for those.</strike> 1.9.3 support was restored, actually.

deivid-rodriguez commented:
Supersedes #4589 and #4510.

What do you think about this? The build is passing but I really don't want to bother installing and fighting against Ruby 1.9.3 & Rails 3, that were EOLed maybe about 2 years ago. So I dropped support for those.

deivid-rodriguez commented:
To compensate, now all the other supported Ruby + Rails combinations are tested in Travis :)

deivid-rodriguez commented:
Just saw #4618, if you really want to keep 1.9.3 & Rails 3.2 support I can reuse the fix in there (or rebase on top of it if it's merged first).

dlackty commented:
Fantastic work!
I wished I could achieve this in #4510 but didn't have enough time to finish it.

deivid-rodriguez commented:
@dlackty I reused your work so thanks for getting it started! :)

deivid-rodriguez commented:
Ok, I think I'm now happy with this so I removed the [WIP].

Let me know how you want to proceed.

timoschilling commented on activeadmin.gemspec:
```diff
@@ -10,10 +10,13 @@ Gem::Specification.new do |s|
   s.test_files    = `git ls-files -- {spec,features}/*`.split("\n")
 
-  s.required_ruby_version = '>= 1.9.3'
+  s.required_ruby_version = '>= 2.0'
```
Why this? We are supporting 1.9 currently

deivid-rodriguez replied:
Because I'm proposing to drop support.

timoschilling commented:
Nice work, but we can't merge this with dropping Ruby 1.9 and Rails 3.2

deivid-rodriguez commented:
Ok, my intention was to not work at all in the Ruby 1.9 + Rails 3.2 specs since I don't want to install or mess with those.

But since there's another PR where I fix is proposed, I'll copy the fix from there and see if it passes Travis.

timoschilling commented:
Please do that! At the moment this PR is grate, but useless

deivid-rodriguez commented:
Sure, I'll get to it! I'll split the changes to drop 1.9 + 3.2 support and leave them in a separate PR, in case you want that sometime in the future.

deivid-rodriguez commented:
I've now restored support for old versions.

I'll now let Travis run and fix issues as they appear. It'll take a while since the matrix now is pretty big (that's a good remainder of the amount of things you are trying to support at the same time).

deivid-rodriguez commented:
Oh well, looks like I'll have to install 1.9.3 after all... :disappointed: 

So what's your plan, release 1.0.0 and then drop support? If yes, why?

timoschilling commented:
Yes we will make a 1.0.0 Release with Ruby 1.9 and Rails 3.2.

We have many users with use one of that dependences and our code base supports both.

zloyrusskiy commented:
Is there way to speed up new release? I'm using Rails 5, and would like to install activeadmin without additional options in Gemfile.

deivid-rodriguez commented:
> Yes we will make a 1.0.0 Release with Ruby 1.9 and Rails 3.2.

The more I investigate the more obscure your version policy looks to me. It seems that you've been _intentionally_ supporting only 1.9 with Rails 3.2 and 4.2 for quite a while. So you basically kept support for the old 3.2 version while breaking it for 4.0 and 4.1 (see #4556 and #4254).

> We have many users with use one of that dependences and our code base supports both. 

Any numbers to backup this? As I said, support for 1.9.3 + (4.0 & 4.1) has been broken for 6 months and only a confused user not aware of being using 1.9.3 and a plugin mantainer have complained. 

Anyways, fixes (including #4556) coming.

deivid-rodriguez commented:
This is ready.

I ended up not removing the `listen` dependency. That dependency comes from `guard` but since the `Guardfile` was last changed back in 2011, I assumed none of the current developers are using it. Given the complexity of the current set of dependencies, I think it's better to remove any unnecessary dependencies to try to prevent similar failures in the future.

Summarizing here the list of related issues for autoclosing on merge:

* Closes #3640.
* Closes #4390.
* Closes #4510.
* Closes #4589.

* Fixes #4369.
* Fixes #4556.

deivid-rodriguez commented:
@timoschilling Rebased on top of current master and removed 2 commits. One was completely unrelated to the PR, the other one had already landed in master through another PR.

You asked me to restore 1.9.3 & Rails 3.2 support and I've done it. Anything else you need here?

deivid-rodriguez commented:
Rebased again, and fixed two new failures because test dependencies dropped ruby 1.9 support.

deivid-rodriguez commented:
Rebased again. Any news @timoschilling?

dmitry commented:
/cc @timoschilling

deivid-rodriguez commented:
Tried rebasing this again and found that the build is again broken because dependencies keep dropping 1.9.3 support. Once maintainers state they want this and they are up for reviewing and merging, I'll fix this.

deivid-rodriguez commented:
Since we now have a new mantainer, @varyonic, and someone else, @Fivell, got master green again. I rebased this PR one more time! :smiley:

deivid-rodriguez commented:
Picked up this PR again. I extracted some (mostly unrelated) things to separate PRs, rebased on top of latest changes in master and improved some stuff.

@varyonic, TravisCI, @timoschilling Please review.

deivid-rodriguez commented:
I'm splitting stuff out of this PR into smaller ones to make it more digestible. So it's not ready yet

deivid-rodriguez commented:
Added 2.4.0 to Travis matrix. Run it locally and got a scary segmentation fault though :S

deivid-rodriguez commented:
This PR was huge so I've been extracting changes to smaller PR's. The idea is to leave this one only for migrating to `Appraisals`. I will ping you when it's ready.

deivid-rodriguez commented:
Ok, this is now ready. This is just one extra change I would like to do which is committing all bundler lock files to source control. This way we ensure that all developers AND CI run against the exact same dependencies so we are sure all environments are consistent and our build doesn't break because of changes in dependencies. What do you think?

deivid-rodriguez commented:
Squashed and lock files added!

timoschilling commented on Gemfile.lock:
```diff
@@ -0,0 +1,210 @@
+GEM
```
I'm not sure that this is a good idea, maybe move this comment to a separate PR.

deivid-rodriguez replied:
@timoschilling What's your concern? This obviously needs more maintenance but I thinks it's definitely less trouble than continuously broken builds because of 3rd party dependency changes. I strongly believe it is a good idea to enforce environment consistency between all developers and CI.

timoschilling replied:
@deivid-rodriguez I agree with you that it brings trouble with 3rd party dependency changes. But this is a feature and not a bug 😉 . If commit the lock would find this issues and often have PRs/commits with dependency updates. Nearly every gem don't check in there Gemfile.lock. Event `bundle gem`'s gem scaffold tells you not do that. I disagree with this change, until some more/better reasons or all other maintainers will agree with it. Please remove it and if you still want a discussion create a new separate PR.

deivid-rodriguez replied:
I can understand that reason although I disagree with it. It think it's a similar debate to using random vs deterministic data for testing (hope you can see the analogy). Support for third party dependencies should be done proactively, it should never get in the middle of unrelated work.

In any case, I understand this is controversial so I'll remove the change!

And by the way, I was really positive you would agree with this because you committed all .lock files to source control when you first tried to add `Appraisal` in #3640. I see now that it was an accident :)

deivid-rodriguez replied:
Example: https://github.com/activerecord-hackery/ransack/pull/755... Some third party gem issue is getting in the middle and now you are forced to either merge a red PR (like I've seen you doing here in the past), hold back the change or start trying to figure out the problem with bundler/third party dependency... no good.

timoschilling replied:
Please make a new PR for a dedicated discussion.

deivid-rodriguez replied:
Commit adding lock files removed! Will open a separate PR once this lands.

deivid-rodriguez commented on .travis.yml:
```diff
@@ -12,7 +12,7 @@ before_install:
   - gem update --system # use the very latest Rubygems
   - rvm @global do gem uninstall bundler -a -x
-  - rvm @global do gem install bundler -v 1.14.2 # latest version known to work
+  - rvm @global do gem install bundler -v 1.14.3 # latest version known to work
```
Well, I introduced this because of the same reason I proposed to commit lock files. https://github.com/activeadmin/activeadmin/pull/4744 was broken because of a buggy bundler release. @varyonic [pointed out](https://github.com/activeadmin/activeadmin/pull/4744#issuecomment-274338610) that it was too bad that our build breaks because of that, so I agreed and locked the version in https://github.com/activeadmin/activeadmin/pull/4744/commits/482734534adb69ca4eb08c889bfb810708f3ac8f.

timoschilling commented:
@deivid-rodriguez please check travis

deivid-rodriguez commented:
Yes, I'm planning to do that after I get merged and rebase on top of my other PR's.

deivid-rodriguez commented:
Ok, I reviewed this. One line had been lost at some point (probably during one of the multiple rebase/commit extractions). Fixed in https://github.com/activeadmin/activeadmin/pull/4624/commits/cee824c340c099552d341eda07b43976121d3f19. I guess you want me to squash that one with https://github.com/activeadmin/activeadmin/pull/4624/commits/4f29ce5481b6f80a909d82b1451311689e186538, right?

timoschilling commented:
please rebase

deivid-rodriguez commented:
Done!

deivid-rodriguez commented:
Nice. There's one last improvement I wanted to add here but I forgot. Make it so that the default `rake` task runs the tests against the latest Rails version. I'll add this as a separate PR if you find it useful.
